enterprise/server/remote_execution/containers/sandbox/sandbox.go

```diff
@@ -0,0 +1,311 @@
+		return "regex"
+	case literalPath:
+		return "literal"
+	case subPath:
+		return "subpath"
+	default:
+		return "unknwon-sbx-path-type"
+	}
+}
+
+func (s *sbxPath) String() string {
+	return fmt.Sprintf(`(%s "%s")`, s.typeString(), s.expression)
+}
+
+func NewSubPath(expr string) sbxPath {
+	return sbxPath{expression: expr, pathType: subPath}
+}
+
+func NewRegexPath(expr string) sbxPath {
+	return sbxPath{expression: expr, pathType: regexPath}
+}
+
+func runSimpleCommand(ctx context.Context, command []string) *interfaces.CommandResult {
+	return commandutil.Run(ctx, &repb.Command{Arguments: command}, "" /*=workDir*/, nil /*=stdin*/, nil /*=stdout*/)
+}
+
+func computeSandboxingSupported(ctx context.Context) bool {
+	if runtime.GOOS != "darwin" {
+		return false
+	}
+	command := []string{
+		sandboxExecBinary,
+		"-p",
+		"(version 1) (allow default)",
+		"/usr/bin/true",
+	}
+	result := runSimpleCommand(ctx, command)
+	return result.ExitCode == 0 && result.Error == nil
+}
+
+func getConfString(ctx context.Context, confVar string) (string, error) {
+	command := []string{getconfBinary, confVar}
+	result := runSimpleCommand(ctx, command)
+	if result.Error != nil {
+		return "", result.Error
+	}
+	return strings.TrimSpace(string(result.Stdout)), nil
+}
+
+func resolveSymlink(path string, fileInfo os.FileInfo) string {
+	if fileInfo.Mode()&os.ModeSymlink != 0 {
+		if absPath, err := os.Readlink(path); err == nil {
+			return absPath
+		}
+	}
+	return path
+}
+
+func resolveAlwaysWritableDirs(ctx context.Context) ([]sbxPath, error) {
+	dirSet := make(map[string]struct{}, 0)
+	dirs := []string{
+		"/dev",
+		"/tmp",
+		"/private/tmp",
+		"/private/var/tmp",
+	}
+
+	// On macOS, processes may write to not only $TMPDIR but also to two
+	// other temporary directories. We have to get their location by calling
+	// "getconf".
+	darwinUserTmp, err := getConfString(ctx, "DARWIN_USER_TEMP_DIR")
+	if err != nil {
+		return nil, err
+	}
+	darwinUserCache, err := getConfString(ctx, "DARWIN_USER_CACHE_DIR")
+	if err != nil {
+		return nil, err
+	}
+	// These directories seem identical to their "/private" prefixed
+	// counterparts, but sandboxed commands will fail to write if they do
+	// not begin with "/private".
+	const privatePrefix = "/private"
+	if !strings.HasPrefix(darwinUserTmp, privatePrefix) {
+		darwinUserTmp = privatePrefix + darwinUserTmp
+	}
+	if !strings.HasPrefix(darwinUserCache, privatePrefix) {
+		darwinUserCache = privatePrefix + darwinUserCache
+	}
+	dirs = append(dirs, darwinUserTmp)
+	dirs = append(dirs, darwinUserCache)
+
+	// Add some user specific dirs as well.
+	homeDir := "/"
+	if h, err := os.UserHomeDir(); err == nil {
+		homeDir = h
+	}
+	dirs = append(dirs, filepath.Join(homeDir, "Library/Caches"))
+	dirs = append(dirs, filepath.Join(homeDir, "Library/Logs"))
+	dirs = append(dirs, filepath.Join(homeDir, "Library/Developer"))
+
+	for _, path := range dirs {
+		if fileInfo, err := os.Lstat(path); !os.IsNotExist(err) {
```

Comment:
nit: I don't understand the logic for checking `!os.IsNotExist(err)` here, maybe add a comment?

Reply:
This is a somewhat confusing API, but this is just checking the file exists:

https://stackoverflow.com/questions/12518876/how-to-check-if-a-file-exists-in-go
(you need to read the whole thing because it contains some wrong answers lol)

Reply:
It seems like `os.IsNotExist(someRandomIOError)` would return `false`, therefore `!os.IsNotExist(err)` would return true, and you would unintentionally execute the `if {}` body

Reply:
I had to write a test to see it but of course you're right :) https://go.dev/play/p/bDmgptfJCly

Sent https://github.com/buildbuddy-io/buildbuddy/pull/1932 to fix.